README.md
OpenStack Boot Camp
====================

Kopos
-----

Kubernetes OPerate OpenStack

### Prerequisites to build *protoc*, *protoc-gen-go*, *protoc-gen-grpc-gateway*

protoc - see also https://github.com/tangfeixiong/go-to-Kubernetes/blob/master/docs/protobuf
```
vagrant@vagrant-ubuntu-trusty-64:/Users/fanhongling/Downloads/workspace/src/github.com/tangfeixiong/go-to-openstack-bootcamp/kopos/echopb$ protoc --version
libprotoc 3.3.0
```

protoc-gen-go
```
vagrant@vagrant-ubuntu-trusty-64:/Users/fanhongling/Downloads/go-kubernetes/src/github.com/golang/protobuf$ go install -v ./protoc-gen-go/
github.com/golang/protobuf/proto
github.com/golang/protobuf/protoc-gen-go/descriptor
github.com/golang/protobuf/protoc-gen-go/plugin
github.com/golang/protobuf/protoc-gen-go/generator
github.com/golang/protobuf/protoc-gen-go/grpc
```

protoc-gen-grpc-gateway
```
vagrant@vagrant-ubuntu-trusty-64:/Users/fanhongling/Downloads/go-kubernetes/src/github.com/grpc-ecosystem/grpc-gateway$ GOPATH=/Users/fanhongling/Downloads/go-kubernetes go build -v -o /home/vagrant/go/bin/protoc-gen-grpc-gateway ./protoc-gen-grpc-gateway/
github.com/golang/glog
github.com/golang/protobuf/proto
github.com/grpc-ecosystem/grpc-gateway/utilities
github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway/httprule
github.com/golang/protobuf/protoc-gen-go/descriptor
github.com/golang/protobuf/protoc-gen-go/plugin
google.golang.org/genproto/googleapis/api/annotations
github.com/golang/protobuf/protoc-gen-go/generator
github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway/descriptor
github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway/generator
github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway/gengateway
github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway
```

protoc-gen-swagger
```
vagrant@vagrant-ubuntu-trusty-64:/Users/fanhongling/Downloads/go-kubernetes/src/github.com/grpc-ecosystem/grpc-gateway$ GOPATH=/Users/fanhongling/Downloads/go-kubernetes go build -v -o /home/vagrant/go/bin/protoc-gen-swagger ./protoc-gen-swagger/
github.com/golang/protobuf/proto
github.com/golang/glog
github.com/grpc-ecosystem/grpc-gateway/utilities
github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway/httprule
github.com/golang/protobuf/protoc-gen-go/descriptor
github.com/golang/protobuf/protoc-gen-go/plugin
google.golang.org/genproto/googleapis/api/annotations
github.com/golang/protobuf/protoc-gen-go/generator
github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway/descriptor
github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway/generator
github.com/grpc-ecosystem/grpc-gateway/protoc-gen-swagger/genswagger
github.com/grpc-ecosystem/grpc-gateway/protoc-gen-swagger
```

### Generate stub of *Protobuf*, *gRPC* and *gRPC-gateway*

Vagrant up (with synced-folder)
```
fanhonglingdeMacBook-Pro:trusty64 fanhongling$ vagrant up
Bringing machine 'default' up with 'virtualbox' provider...
==> default: Checking if box 'ubuntu/trusty64' is up to date...
==> default: A newer version of the box 'ubuntu/trusty64' is available! You currently
==> default: have version '20160122.0.0'. The latest is version '20170602.0.0'. Run
==> default: `vagrant box update` to update.
==> default: Clearing any previously set forwarded ports...
==> default: Clearing any previously set network interfaces...
==> default: Preparing network interfaces based on configuration...
    default: Adapter 1: nat
    default: Adapter 2: hostonly
==> default: Forwarding ports...
    default: 6443 (guest) => 6443 (host) (adapter 1)
    default: 22 (guest) => 2222 (host) (adapter 1)
==> default: Running 'pre-boot' VM customizations...
==> default: Booting VM...
==> default: Waiting for machine to boot. This may take a few minutes...
    default: SSH address: 127.0.0.1:2222
    default: SSH username: vagrant
    default: SSH auth method: private key
    default: Warning: Remote connection disconnect. Retrying...
    default: Warning: Remote connection disconnect. Retrying...
==> default: Machine booted and ready!
==> default: Checking for guest additions in VM...
==> default: Configuring and enabling network interfaces...
==> default: Mounting shared folders...
    default: /vagrant => /Users/fanhongling/https%3A%2F%2Fwww.vagrantup.com/https%3A%2F%2Fatlas.hashicorp.com/boxes/ubuntu/trusty64
    default: /Users/fanhongling/go/src => /Users/fanhongling/go/src
    default: /Users/fanhongling/go/pkg => /Users/fanhongling/go/pkg
    default: /Users/fanhongling/Downloads/workspace => /Users/fanhongling/Downloads/workspace
    default: /Users/fanhongling/Downloads/99-mirror => /Users/fanhongling/Downloads/99-mirror
    default: /Users/fanhongling/Downloads/go-openshift/pkg => /Users/fanhongling/Downloads/go-openshift/pkg
    default: /Users/fanhongling/Downloads/go-openshift/src => /Users/fanhongling/Downloads/go-openshift/src
    default: /Users/fanhongling/Downloads/go-kubernetes/src => /Users/fanhongling/Downloads/go-kubernetes/src
    default: /Users/fanhongling/Downloads/go-kubernetes/pkg => /Users/fanhongling/Downloads/go-kubernetes/pkg
==> default: Machine already provisioned. Run `vagrant provision` or use the `--provision`
==> default: flag to force provisioning. Provisioners marked to run always will still run.
fanhonglingdeMacBook-Pro:trusty64 fanhongling$ ssh vagrant@172.17.4.200
vagrant@172.17.4.200's password: 
Welcome to Ubuntu 14.04.5 LTS (GNU/Linux 3.13.0-76-generic x86_64)

 * Documentation:  https://help.ubuntu.com/

  System information as of Wed Jun  7 12:40:07 UTC 2017

  System load:  0.55               Users logged in:        0
  Usage of /:   93.0% of 39.34GB   IP address for eth0:    10.0.2.15
  Memory usage: 3%                 IP address for eth1:    172.17.4.200
  Swap usage:   0%                 IP address for docker0: 172.18.0.1
  Processes:    120

  => / is using 93.0% of 39.34GB

  Graph this data and manage this system at:
    https://landscape.canonical.com/

  Get cloud support with Ubuntu Advantage Cloud Guest:
    http://www.ubuntu.com/business/services/cloud

51 packages can be updated.
31 updates are security updates.

New release '16.04.2 LTS' available.
Run 'do-release-upgrade' to upgrade to it.


Last login: Sun Jun  4 06:27:37 2017 from 172.17.4.1
vagrant@vagrant-ubuntu-trusty-64:~$ cd /Users/fanhongling/Downloads/workspace/src/github.com/tangfeixiong/go-to-openstack-bootcamp/kopos/echopb/
```

Protocol buffer
```
vagrant@vagrant-ubuntu-trusty-64:/Users/fanhongling/Downloads/workspace/src/github.com/tangfeixiong/go-to-openstack-bootcamp/kopos/echopb$ make openstack/neutron
protoc -I/usr/local/include -I. \
		-I/Users/fanhongling/Downloads/workspace/src \
		-I/Users/fanhongling/Downloads/workspace/src/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis \
		--go_out=plugins=grpc:. \
		openstack/neutron/neutron.proto
go generate .
```

For gRPC and gRPC-gateway
```
vagrant@vagrant-ubuntu-trusty-64:/Users/fanhongling/Downloads/workspace/src/github.com/tangfeixiong/go-to-openstack-bootcamp/kopos/echopb$ make all
protoc -I/usr/local/include -I. \
		-I/Users/fanhongling/Downloads/go-kubernetes:/Users/fanhongling/Downloads/go-openshift:/Users/fanhongling/Downloads/workspace/src \
		-I/Users/fanhongling/Downloads/go-kubernetes:/Users/fanhongling/Downloads/go-openshift:/Users/fanhongling/Downloads/workspace/src/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis \
		--go_out=plugins=grpc:. \
		service.proto data.proto
protoc -I/usr/local/include -I. \
		-I/Users/fanhongling/Downloads/go-kubernetes:/Users/fanhongling/Downloads/go-openshift:/Users/fanhongling/Downloads/workspace/src \
		-I/Users/fanhongling/Downloads/go-kubernetes:/Users/fanhongling/Downloads/go-openshift:/Users/fanhongling/Downloads/workspace/src/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis \
		--grpc-gateway_out=logtostderr=true:. \
		service.proto
protoc -I/usr/local/include -I. \
		-I/Users/fanhongling/Downloads/go-kubernetes:/Users/fanhongling/Downloads/go-openshift:/Users/fanhongling/Downloads/workspace/src \
		-I/Users/fanhongling/Downloads/go-kubernetes:/Users/fanhongling/Downloads/go-openshift:/Users/fanhongling/Downloads/workspace/src/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis \
		--swagger_out=logtostderr=true:. \
		service.proto
go generate .
```

### Build and Run

Build
```
fanhonglingdeMacBook-Pro:go-to-openstack-bootcamp fanhongling$ go build -o bin/kopos -v ./kopos/
golang.org/x/net/context
github.com/grpc-ecosystem/grpc-gateway/utilities
github.com/golang/protobuf/proto
google.golang.org/grpc/codes
google.golang.org/grpc/grpclog
google.golang.org/grpc/metadata
github.com/spf13/pflag
golang.org/x/sys/unix
github.com/magiconair/properties
gopkg.in/yaml.v2
github.com/fsnotify/fsnotify
golang.org/x/net/http2/hpack
github.com/spf13/cobra
golang.org/x/net/idna
golang.org/x/net/internal/timeseries
golang.org/x/net/lex/httplex
golang.org/x/net/http2
golang.org/x/net/trace
github.com/golang/protobuf/ptypes/struct
github.com/golang/protobuf/jsonpb
github.com/grpc-ecosystem/grpc-gateway/runtime/internal
github.com/golang/protobuf/ptypes/any
google.golang.org/genproto/googleapis/rpc/status
google.golang.org/grpc/status
github.com/spf13/viper
github.com/tangfeixiong/go-to-openstack-bootcamp/kopos/echopb/openstack/neutron
github.com/golang/protobuf/protoc-gen-go/descriptor
github.com/grpc-ecosystem/grpc-gateway/runtime
google.golang.org/genproto/googleapis/api/annotations
google.golang.org/grpc/credentials
google.golang.org/grpc/grpclb/grpc_lb_v1
google.golang.org/grpc/internal
google.golang.org/grpc/keepalive
google.golang.org/grpc/naming
google.golang.org/grpc/peer
google.golang.org/grpc/stats
google.golang.org/grpc/tap
github.com/tangfeixiong/go-to-openstack-bootcamp/kopos/insecure
github.com/tangfeixiong/go-to-openstack-bootcamp/kopos/pkg/ui/data/swagger
google.golang.org/grpc/transport
google.golang.org/grpc
github.com/tangfeixiong/go-to-openstack-bootcamp/kopos/echopb
github.com/tangfeixiong/go-to-openstack-bootcamp/kopos/cmd
github.com/tangfeixiong/go-to-openstack-bootcamp/kopos
```

Server
```
fanhonglingdeMacBook-Pro:go-to-openstack-bootcamp fanhongling$ bin/kopos serve
2017/06/04 03:36:19 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp [::1]:10000: getsockopt: connection refused"; Reconnecting to {localhost:10000 <nil>}
grpc on port: 10000
```

Client test
```
fanhonglingdeMacBook-Pro:go-to-openstack-bootcamp fanhongling$ bin/kopos test create
test create
0xc420176af0
```

OpenStack DevOps
----------------

[Markdown](https://daringfireball.net/projects/markdown/)

[Sphinx - Python Documentation Generator](http://sphinx-doc.org/contents.html)

### OpenStack-Grizzly版在Ubuntu-Precise即12.04.3-LTS的安装实验-

参考[mseknibilel](https://github.com/mseknibilel/OpenStack-Grizzly-Install-Guide/tree/OVS_SingleNode)的安装指南，向作者致谢

#### 准备

1. 切换到超级用户root
    sudo su或sudo -i
su是linux命令，Switch User的意思，但当前用户没有执行该命令的权限
sudo命令即Super User Do，使当前用户可以执行root命令

2. 添加Grizzly的apt-get在线安装仓库
apt-get的在线安装仓库配置文件在/etc/apt-get目录下
source.list配置文件是Ubuntu系统的在线更新或升级配置文件
使用国内镜像mirrors.163.com
而第三方的在线更新配置文件可以放置在/etc/apt/source.list.d/目录下
这里将OpenStack的配置文件取名为grizzly.list
可以用vi或vim来创建和编辑
cd /etc/apt/source.list.d/
vi grizzly.list
按insert键
输入deb http://ubuntu-cloud.archive.canonical.com/ubuntu precise-updates/grizzly main
按escape键
按wq键
mseknibilel的文章中采用了重定向stdout到grizzly.list文件的写法：
echo deb http://ubuntu-cloud.archive.canonical.com/ubuntu precise-updates/grizzly main >> /etc/apt/sources.list.d/grizzly.list

3. 添加在Ubuntu Precise上安装OpenStack必要的keyring和python等工具
    apt-get install ubuntu-cloud-keyring python-software-properties software-properties-common python-keyring

4. 更新Ubuntu
    apt-get update
    apt-get upgrade
    apt-get dist-upgrade


kopos/CHANGELOG.md
# API instruction

__OpenAPI__

Swagger 2.0

swagger-ui为https://localhost:10000/swagger-ui

swagger.json为https://localhost:10000/swagger.json

__Table of Contents__

1. [Batch create Virtual Machines](#Batch create Virtual Machines)
2. [Batch create simple Networks](#Batch create simple Networks)
3. [List flavors](#Flavor list)
3. [List images](#Image list)
3. [Get flavor](#Flavor get)
3. [Get image](#Image get)
4. [List networks](#Network list)
4. [List subnetes](#Subnet list)
4. [List VMs](#Virtual machine list)
4. [Reboot VMs](#Reboot virtual machines)
4. [Destroy VMs](#Destroy virtual machines)
5. [Spawn VMs](#Spawn Machines)
5. [Networking Architecture](#Networking Topology)

## V1

API path prefix  _/v1/..._


### 0.1.2-alpha3

2012-07-24

Fix bug  /v1/spawn   with correct json
```
  {
    vms: [
      {
        "flavor_name": "m1.small",
        "image_name": "cirros",
        "min_count": 2,
        "max_count": 4,
        "secgroups_info": [],
        "user_data": [],
        "network_name": "private",
        "floating_network_name": "public",
        "personality": [],
        "name_prefix": "awesome-VM"
      },
      {
        ...
      }
    ]
  }
```

### 0.1.2-alpha2

2012-07-12

### Openstack compute config

1）因获取libvirt的vncdisplay需要，对现用openstack的compute node的配置需显式配置（CHANGELOG）
如在10.121.198.4环境上
```
[root@compute01 ~]# head /etc/nova/nova.conf
[default]
my_ip=10.121.198.4
```

请devops务必在今后，或其它openstack环境中添加my_ip配置项，如10.100.151.149，配置my_ip＝10.100.151.149

检查配置
```
[vagrant@localhost kopos]$ nova hypervisor-show compute01
+---------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Property                  | Value                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
+---------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| cpu_info_arch             | x86_64                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
| cpu_info_features         | ["ssse3", "pge", "avx", "clflush", "sep", "syscall", "vme", "dtes64", "invpcid", "msr", "sse", "xsave", "vmx", "erms", "xtpr", "cmov", "tsc", "smep", "pbe", "est", "pat", "monitor", "smx", "lm", "abm", "nx", "fxsr", "tm", "sse4.1", "pae", "sse4.2", "pclmuldq", "acpi", "fma", "tsc-deadline", "popcnt", "mmx", "osxsave", "cx8", "mce", "mtrr", "rdtscp", "ht", "pse", "lahf_lm", "pdcm", "mca", "pdpe1gb", "apic", "fsgsbase", "f16c", "ds", "invtsc", "pni", "tm2", "avx2", "aes", "sse2", "ss", "bmi1", "bmi2", "pcid", "de", "fpu", "cx16", "pse36", "ds_cpl", "movbe", "rdrand", "x2apic"] |
| cpu_info_model            | SandyBridge                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
| cpu_info_topology_cores   | 4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |

kopos/cmd/kopos.go
// tangfeixiong <tangfx128@gmail.com>

package cmd

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/spf13/cobra"
	"github.com/spf13/pflag"
	"golang.org/x/net/context"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
	"google.golang.org/grpc/grpclog"
	"k8s.io/kubernetes/pkg/util/rand"

	"github.com/tangfeixiong/go-to-openstack-bootcamp/kopos/cmd/sapcc-koposcmd"
	pb "github.com/tangfeixiong/go-to-openstack-bootcamp/kopos/echopb"
	pbos "github.com/tangfeixiong/go-to-openstack-bootcamp/kopos/echopb/openstack"
)

// testCmd represents the test command
var testCmd = &cobra.Command{
	Use:   "test",
	Short: "Example test gRPC service CLI client",
	Run: func(cmd *cobra.Command, args []string) {
		var insecure bool
		pflag.BoolVar(&insecure, "insecure", true, "using http.")
		pflag.Parse()
		if insecure {
			fmt.Println("http...")
			url := "http://localhost:10001/v1/battlefields"

			var netTransport = &http.Transport{
				Dial: (&net.Dialer{
					Timeout: 5 * time.Second,
				}).Dial,
				// TLSHandshakeTimeout: 5 * time.Second,
			}
			var netClient = &http.Client{
				Timeout:   time.Second * 10,
				Transport: netTransport,
			}

			in, err := json.Marshal(&pbos.OpenstackNeutronNetRequestData{Name: "test"})
			if err != nil {
				panic(err)
			}

			response, err := netClient.Post(url, "application/json", bytes.NewBuffer(in))
			if err != nil {
				panic(err)
			}
			defer response.Body.Close()
			fmt.Println("response Status:", response.Status)
			fmt.Println("response Headers:", response.Header)
			respbody, _ := ioutil.ReadAll(response.Body)
			fmt.Println("response Body:", string(respbody))

			var jsonStr = []byte(`{"name": "again"}`)
			req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonStr))
			req.Header.Set("X-Custom-Header", "test again")
			req.Header.Set("Content-Type", "application/json")

			client := &http.Client{}
			resp, err := client.Do(req)
			if err != nil {
				panic(err)
			}
			defer resp.Body.Close()

			fmt.Println("response Status:", resp.Status)
			fmt.Println("response Headers:", resp.Header)
			body, _ := ioutil.ReadAll(resp.Body)
			fmt.Println("response Body:", string(body))

			return
		}

		println("grpc with tls")
		var opts []grpc.DialOption
		creds := credentials.NewClientTLSFromCert(demoCertPool, "localhost:10000")
		opts = append(opts, grpc.WithTransportCredentials(creds))
		conn, err := grpc.Dial(demoAddr, opts...)
		if err != nil {
			grpclog.Fatalf("fail to dial: %v", err)
		}
		defer conn.Close()
		client := pb.NewEchoServiceClient(conn)

		// msg, err := client.Echo(context.Background(), &pb.EchoMessage{strings.Join(os.Args[2:], " ")})
		// println(msg.Value)
		println(strings.Join(os.Args[1:], " "))

		in := &pbos.OpenstackNeutronNetRequestData{Name: "test"}
		// copts := []grpc.CallOption{grpc.EmptyCallOption{}}
		copts := []grpc.CallOption{}
		resp, err := client.AdminSharedNetworkCreation(context.Background(), in, copts...)
		println(resp)
	},
}

func init() {
	testCmd.AddCommand(koposcmd.SapccKoposCmd)
	testCmd.AddCommand(tokenCommand, bootCommand(), netCommand())
	RootCmd.AddCommand(testCmd)

	// Here you will define your flags and configuration settings.

	// Cobra supports Persistent Flags which will work for this command
	// and all subcommands, e.g.:
	// echoCmd.PersistentFlags().String("foo", "", "A help for foo")

	// Cobra supports local flags which will only run when this command
	// is called directly, e.g.:
	// echoCmd.Flags().String("foo", "", "A help for foo")

	testCmd.PersistentFlags().BoolP("insecure", "n", true, "using insecure rpc/http")
	tokenCommand.Flags().BoolP("http", "t", true, "using insecure http")
}

var tokenCommand = &cobra.Command{
	Use:   "token",
	Short: "Example token gRPC service CLI client",
	Run: func(cmd *cobra.Command, args []string) {
		var insecure, ishttp bool
		pflag.BoolVar(&insecure, "insecure", true, "using insecure rpc/http.")
		pflag.BoolVar(&ishttp, "http", true, "using http.")
		pflag.Parse()

		if insecure {
			if ishttp {
				var url = "http://localhost:10001/v1/token"
				var jsonStr = []byte(`{"src": null, "tgt": null}`)
				req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonStr))
				req.Header.Set("X-Custom-Header", "test again")
				req.Header.Set("Content-Type", "application/json")

				client := &http.Client{}
				resp, err := client.Do(req)
				if err != nil {
					panic(err)
				}
				defer resp.Body.Close()

				fmt.Println("response Status:", resp.Status)
				fmt.Println("response Headers:", resp.Header)
				body, _ := ioutil.ReadAll(resp.Body)
				fmt.Println("response Body:", string(body))
				return
			}

			conn, err := grpc.Dial("localhost:10002", grpc.WithInsecure())
			if err != nil {
				log.Fatalf("did not connect: %v", err)
			}
			defer conn.Close()
			c := pb.NewEchoServiceClient(conn)

			// Contact the server and print out its response.
			if len(os.Args) > 1 {
				log.Printf("args: %+v", os.Args[1:])
			}
			r, err := c.ValidateToken(context.Background(), &pbos.TokenReqRespData{})
			if err != nil {
				log.Fatalf("could not greet: %v", err)
			}
			log.Printf("Greeting: %v", r)
			return
		}

		var opts []grpc.DialOption
		creds := credentials.NewClientTLSFromCert(demoCertPool, "localhost:10000")
		opts = append(opts, grpc.WithTransportCredentials(creds))
		conn, err := grpc.Dial(demoAddr, opts...)
		if err != nil {
			grpclog.Fatalf("fail to dial: %v", err)
		}
		defer conn.Close()
		client := pb.NewEchoServiceClient(conn)

		println(strings.Join(os.Args[1:], " "))

		in := &pbos.TokenReqRespData{}
		// copts := []grpc.CallOption{grpc.EmptyCallOption{}}
		copts := []grpc.CallOption{}
		resp, err := client.ValidateToken(context.Background(), in, copts...)
		if err != nil {
			grpclog.Printf("fail to greet: %v", err)
		}
		println(resp)

	},
}

func bootCommand() *cobra.Command {
	var insecure, ishttp bool
	var multimode bool
	bootcmd := &cobra.Command{
		Use:   "boot",
		Short: "Example boot gRPC service CLI client",
		Run: func(cmd *cobra.Command, args []string) {
			pflag.BoolVarP(&insecure, "insecure", "n", true, "using insecure rpc/http.")
			pflag.BoolVarP(&ishttp, "http", "t", true, "using http.")
			pflag.BoolVar(&multimode, "multimode", false, "multiple flavor and image")
			pflag.Parse()

			if insecure {
				if ishttp {
					var url = "http://localhost:10001/v1/boot"
					var jsonStr = []byte(strings.Replace(`
{
  "flavor_name": "m1.tiny",
  "image_name": "cirros",
  "min_count": 1,
  "max_count": 2,
  "secgroups_info": [
    {
      "id": "15895b59-9aa7-4e6e-9d23-75a9324030f5",
      "name": "default"
    }
  ],
  "user_data": [],
  "network_name": "private",
  "floating_network_name": "public-192-168-1-0",
  "personality": [],
  "name_prefix": "............"
}
`, "............", rand.String(12), 1))
					if multimode {
						jsonStr = []byte(strings.Replace(`
{ "vms": [{
  "flavor_name": "m1.tiny",
  "image_name": "cirros",
  "min_count": 1,
  "max_count": 2,
  "secgroups_info": [
    {
      "id": "15895b59-9aa7-4e6e-9d23-75a9324030f5",
      "name": "default"
    }
  ],
  "user_data": [],
  "network_name": "private",
  "floating_network_name": "public-192-168-1-0",
  "personality": [],
  "name_prefix": "............"
},
{
  "flavor_name": "m1.tiny",
  "image_name": "cirros",
  "min_count": 1,
  "max_count": 2,
  "secgroups_info": [
    {
      "id": "15895b59-9aa7-4e6e-9d23-75a9324030f5",
      "name": "default"
    }
  ],
  "user_data": [],
  "network_name": "private",
  "floating_network_name": "public-192-168-1-0",
  "personality": [],
  "name_prefix": "............"
}]}`, "............", rand.String(12), 1))
						url = "http://localhost:10001/v1/spawn"
					}
					req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonStr))
					if err != nil {
						panic(err)
					}
					req.Header.Set("X-Custom-Header", "test again")
					req.Header.Set("Content-Type", "application/json")

					client := &http.Client{}
					resp, err := client.Do(req)
					if err != nil {
						panic(err)
					}
					defer resp.Body.Close()

					fmt.Println("response Status:", resp.Status)
					fmt.Println("response Headers:", resp.Header)
					body, _ := ioutil.ReadAll(resp.Body)
					fmt.Println("response Body:", string(body))
					return
				}

				conn, err := grpc.Dial("localhost:10002", grpc.WithInsecure())
				if err != nil {
					log.Fatalf("did not connect: %v", err)
				}
				defer conn.Close()
				c := pb.NewEchoServiceClient(conn)

				// Contact the server and print out its response.
				if len(os.Args) > 1 {
					log.Printf("args: %+v", os.Args[1:])
				}
				r, err := c.ValidateToken(context.Background(), &pbos.TokenReqRespData{})
				if err != nil {
					log.Fatalf("could not greet: %v", err)
				}
				log.Printf("Greeting: %v", r)
				return
			}

			var opts []grpc.DialOption
			creds := credentials.NewClientTLSFromCert(demoCertPool, "localhost:10000")
			opts = append(opts, grpc.WithTransportCredentials(creds))
			conn, err := grpc.Dial(demoAddr, opts...)
			if err != nil {
				grpclog.Fatalf("fail to dial: %v", err)
			}
			defer conn.Close()
			client := pb.NewEchoServiceClient(conn)

			println(strings.Join(os.Args[1:], " "))

			in := &pbos.TokenReqRespData{}
			// copts := []grpc.CallOption{grpc.EmptyCallOption{}}
			copts := []grpc.CallOption{}
			resp, err := client.ValidateToken(context.Background(), in, copts...)
			if err != nil {
				grpclog.Printf("fail to greet: %v", err)
			}
			println(resp)

		},
	}

	bootcmd.Flags().BoolVarP(&ishttp, "http", "t", true, "using insecure http.")
	bootcmd.Flags().BoolVar(&multimode, "multimode", false, "multiple flavor and image")

	bootcmd.Flags().StringP("name", "", "", "vm name and/or prefix")
	bootcmd.Flags().StringP("flavor_name", "", "m1.small", "vm spec")
	bootcmd.Flags().StringP("image_name", "", "cirros", "vm image")
	bootcmd.Flags().IntP("count", "", 4, "vm count")
	bootcmd.Flags().StringP("secgroup_names", "", "default", "security group name array seperated with comma")
	bootcmd.Flags().StringP("network_name", "", "", "vm network name")
	bootcmd.Flags().StringP("floating_network_name", "", "", "vm floating ip network name")
	bootcmd.Flags().StringP("userdata_file", "", "", "user data file")
	bootcmd.Flags().StringP("personality_files", "", "", "personality file array seperated with comma")

	return bootcmd
}

func netCommand() *cobra.Command {
	netcmd := &cobra.Command{
		Use:   "net",
		Short: "Example networking gRPC service CLI client",
		Run: func(cmd *cobra.Command, args []string) {
			var insecure, ishttp bool
			pflag.BoolVar(&insecure, "insecure", true, "using insecure rpc/http.")
			pflag.BoolVar(&ishttp, "http", true, "using http.")
			pflag.Parse()

			if insecure {
				if ishttp {
					var url = "http://localhost:10001/v1/landscape"
					var jsonStr = []byte(strings.Replace(`
{
  "vnets": [
    {
      "name": "int-stage-0",
      "subnets": [
        {
          "name": "int-192-168-128-0-slash-24",
          "cidr": "192.168.128.0/24",
          "dns_name_servers": [],
          "allocation_pools": [],
          "host_routes": [],
          "enable_dhcp": true
        }
      ]
    },
    {
      "name": "int-stage-1",
      "subnets": [
        {
          "name": "int-192-168-129-0-slash-24",
          "cidr": "192.168.129.0/24",
          "dns_name_servers": [],
          "allocation_pools": [],
          "host_routes": [],
          "enable_dhcp": true
        }
      ]
    },
    {
      "name": "int-stage-2",
      "subnets": [
        {
          "name": "int-192-168-130-0-slash-24",
          "cidr": "192.168.130.0/24",
          "dns_name_servers": [],
          "allocation_pools": [],
          "host_routes": [],
          "enable_dhcp": true
        }
      ]
    },
    {
      "name": "int-stage-3",
      "subnets": [
        {
          "name": "int-192-168-131-0-slash-24",
          "cidr": "192.168.131.0/24",
          "dns_name_servers": [],
          "allocation_pools": [],
          "host_routes": [],
          "enable_dhcp": true
        }
      ]
    },
    {
      "name": "public",
      "admin_state_up": true,
      "subnets": [
        {
          "name": "10.100.151.0/24",
          "cidr": "10.100.151.0/24",
          "gateway_ip": "10.100.151.1",
          "dns_name_servers": [],
          "allocation_pools": [
            {
              "start": "10.100.151.50",
              "end": "10.100.151.240"
            }
          ],
          "host_routes": [],
          "enable_dhcp": false
        }
      ],
      "shared": true
    }
  ],
  "vrouter": {
    "name": "hack",
    "routes": [],
    "admint_state_up": true
  },
  "secgroup": {
    "name": "hack",
    "security_group_rules": [
      {
        "direction": "ingress",
        "protocol": "tcp"
      },
      {
        "direction": "ingress",
        "protocol": "udp"
      },
      {
        "direction": "ingress",
        "protocol": "icmp"
      }
    ]
  },
  "ifaces_info": [
    {
      "router_name": "hack",
      "network_name": "int-stage-0",
      "subnet_name": "int-192-168-128-0-slash-24",
      "secgroups_info": [
        {
          "name": "hack"
        }
      ]
    },
    {
      "router_name": "hack",
      "network_name": "int-stage-1",
      "subnet_name": "int-192-168-129-0-slash-24",
      "secgroups_info": [
        {
          "name": "hack"
        }
      ]
    },
    {
      "router_name": "hack",
      "network_name": "int-stage-2",
      "subnet_name": "int-192-168-130-0-slash-24",
      "secgroups_info": [
        {
          "name": "hack"
        }
      ]
    },
    {
      "router_name": "hack",
      "network_name": "int-stage-3",
      "subnet_name": "int-192-168-131-0-slash-24",
      "secgroups_info": [
        {
          "name": "hack"
        }
      ]
    }
  ],
  "gateways_info": [
    {
      "network_name": "public",
      "router_name": "hack"
    }
  ]
}
`, "............", rand.String(12), 1))
					obj := new(pbos.OpenstackNeutronLandscapeReqRespData)
					if err := json.Unmarshal(jsonStr, obj); err != nil {
						log.Fatal(err)
					} else {
						log.Println(*obj)
					}
					req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonStr))
					req.Header.Set("X-Custom-Header", "test again")
					req.Header.Set("Content-Type", "application/json")

					client := &http.Client{}
					resp, err := client.Do(req)
					if err != nil {
						panic(err)
					}
					defer resp.Body.Close()

					fmt.Println("response Status:", resp.Status)
					fmt.Println("response Headers:", resp.Header)
					body, _ := ioutil.ReadAll(resp.Body)
					fmt.Println("response Body:", string(body))
					return
				}

				conn, err := grpc.Dial("localhost:10002", grpc.WithInsecure())
				if err != nil {
					log.Fatalf("did not connect: %v", err)
				}
				defer conn.Close()
				c := pb.NewEchoServiceClient(conn)

				// Contact the server and print out its response.
				if len(os.Args) > 1 {
					log.Printf("args: %+v", os.Args[1:])
				}
				r, err := c.EstablishNetworkLandscape(context.Background(), &pbos.OpenstackNeutronLandscapeReqRespData{})
				if err != nil {
					log.Fatalf("could not greet: %v", err)
				}
				log.Printf("Greeting: %v", r)
				return
			}

			var opts []grpc.DialOption
			creds := credentials.NewClientTLSFromCert(demoCertPool, "localhost:10000")
			opts = append(opts, grpc.WithTransportCredentials(creds))
			conn, err := grpc.Dial(demoAddr, opts...)
			if err != nil {
				grpclog.Fatalf("fail to dial: %v", err)
			}
			defer conn.Close()
			client := pb.NewEchoServiceClient(conn)

			println(strings.Join(os.Args[1:], " "))

			in := &pbos.OpenstackNeutronLandscapeReqRespData{}
			// copts := []grpc.CallOption{grpc.EmptyCallOption{}}
			copts := []grpc.CallOption{}
			resp, err := client.EstablishNetworkLandscape(context.Background(), in, copts...)
			if err != nil {
				grpclog.Printf("fail to greet: %v", err)
			}
			println(resp)

		},
	}
	netcmd.Flags().StringP("vnets", "", "", "json expr")
	netcmd.Flags().StringP("vrouter", "", "", "json expr")
	netcmd.Flags().StringP("secgroup", "", "", "json expr")
	netcmd.Flags().StringP("ifaces_info", "", "", "json expr")
	netcmd.Flags().StringP("gateways_info", "", "", "json expr")

	return netcmd
}
